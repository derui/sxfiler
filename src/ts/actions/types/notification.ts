// actions for notification

enum ActionTypes {
  timeout = "notification_timeout",
}

export default ActionTypes;
