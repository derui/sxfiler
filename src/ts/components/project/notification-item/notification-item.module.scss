@import "../../ui/commom-style/_colors.scss";
@import "../../ui/commom-style/_variables.scss";
@import "../../ui/commom-style/reset.module.css";

$messageWidth: 150px;

$borderWidth: 3px;

.root {
  composes: listItem-base from "../../ui/list-item/list-item.module.scss";

  padding: $baseSpace ($baseSpace * 2);
  margin-right: $smallSpace;
  margin-bottom: $smallSpace;

  &[data-level="info"] {
    border: $borderWidth solid darken($sol_blue, 20%);
    background-color: lighten($sol_blue, 20%);
    color: $sol_base02;
  }

  &[data-level="warning"]{

  border: $borderWidth solid darken($sol_orange, 20%);
  background-color: lighten($sol_orange, 20%);
  color: $sol_base03;
}

  &[data-level="error"] {
    border: $borderWidth solid darken($sol_red, 30%);
    background-color: $sol_red;
    color: $sol_base2;
  }
}

.animation{
  /*!*/
}

.animation-enter {
  opacity: 0.01;
}

.animation-enter-active {
  opacity: 1;
  transition: opacity 200ms ease-out;
}

.animation-exit {
  opacity: 1;
}

.animation-exit-active {
  opacity: 0.01;
  transition: opacity 200ms ease-out;
}
