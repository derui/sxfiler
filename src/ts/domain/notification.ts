
// type of notification
export enum NotificationType {
  OneShot = 'oneshot',
  Progress = 'progress',
}

export enum Level {
  Info = 'info',
  Warning = 'warning',
  Error = 'error'
}

export interface OneShotBody {
  kind: "oneshot";
  message: string;
}

export interface ProgressBody {
  kind: "progress";
  process: string;
  current: number;
  target: number;
}

export type Body = OneShotBody | ProgressBody

// notification from server
export class Notification {
  constructor(
    public readonly id: string,
    public readonly level: Level,
    public readonly body: Body
  ) {}
}
