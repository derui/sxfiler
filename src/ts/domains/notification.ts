// type of notification
export enum NotificationType {
  Message = "message",
  Progress = "progress",
}

export enum Level {
  Info = "info",
  Warning = "warning",
  Error = "error",
}

export interface MessageBody {
  kind: "message";
  message: string;
}

export interface ProgressBody {
  kind: "progress";
  process: string;
  current: number;
  target: number;
}

export type Body = MessageBody | ProgressBody;

/**
 * factory function to create one-shot notification
 * @param id
 * @param level
 * @param message
 */
export function createMessage(id: string, level: Level, message: string) {
  return new Notification(id, level, { kind: NotificationType.Message, message });
}

/**
 * factory function to create progress notification
 * @param id
 * @param level
 * @param progress
 */
export function createProgress(
  id: string,
  level: Level,
  progress: { process: string; current: number; target: number }
) {
  return new Notification(id, level, { kind: NotificationType.Progress, ...progress });
}

// notification from server
export class Notification {
  constructor(public readonly id: string, public readonly level: Level, private body: Body) {}

  public get bodyKind() {
    return this.body.kind;
  }

  // get body as progress.
  public getProgressBody(): ProgressBody {
    switch (this.body.kind) {
      case "message":
        throw Error("can not convert to progress body");
      case "progress":
        return this.body;
    }
  }

  // get body as message.
  public getMessageBody(): MessageBody {
    switch (this.body.kind) {
      case "message":
        return this.body;
      case "progress":
        throw Error("can not convert to one-shot body");
    }
  }
}
