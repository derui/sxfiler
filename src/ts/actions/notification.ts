import { Notification } from "../domains/notification";
import { AppAction } from "./type";

export enum ActionTypes {
  timeout = "notification_timeout",
  notify = "notification_notify",
  remove = "notification_remove",
}

type TimeoutAction = AppAction<ActionTypes.timeout, { notificationId: string }>;
type NotifyAction = AppAction<ActionTypes.notify, { notification: Notification }>;
type RemoveAction = AppAction<ActionTypes.remove, { notificationId: string }>;

export type Actions = TimeoutAction | NotifyAction | RemoveAction;

const timeout = (notificationId: string): TimeoutAction => {
  return { type: ActionTypes.timeout, notificationId };
};

const notify = (notification: Notification): NotifyAction => {
  return { type: ActionTypes.notify, notification };
};

const remove = (notificationId: string): RemoveAction => {
  return { type: ActionTypes.remove, notificationId };
};

export const actions = { timeout, notify, remove };
