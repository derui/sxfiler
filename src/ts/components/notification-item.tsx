import classNames from "classnames";
import * as React from "react";
import { CSSTransition } from "react-transition-group";
import { Level, Notification } from "../domain/notification";

export type TimeoutCallback = (id: string) => void;

interface Prop {
  item: Notification;
  timeouted: boolean;
  onItemTimeouted: TimeoutCallback;
}

interface State {
  mounted: boolean;
}

function assertNever(x: never): never {
  throw new Error("Unexpected object: " + x);
}

export default class NotificationItem extends React.Component<Prop, State> {
  constructor(props: Prop) {
    super(props);

    this.state = { mounted: false };
  }

  public componentDidMount() {
    this.setState({ mounted: false });
  }

  public handleItemTimeouted() {
    this.props.onItemTimeouted(this.props.item);
  }

  public render() {
    const { item } = this.props;

    const renderer = () => {
      const className = classNames("fp-NotificationList_Item", {
        "fp-NotificationList_Item-info": item.level === Level.Info,
        "fp-NotificationList_Item-warning": item.level === Level.Warning,
        "fp-NotificationList_Item-error": item.level === Level.Error,
      });

      switch (item.body.kind) {
        case "oneshot":
          return <li className={className}>{item.body.message}</li>;
        case "progress":
          return null;
        default:
          return assertNever(item.body);
      }
    };

    return (
      <CSSTransition
        in={this.state.mounted && this.props.timeouted}
        onExited={this.onItemTimeouted}
        timeout={200}
        classNames="fp-NotificationList_ItemAnimation"
      >
        {renderer}
      </CSSTransition>
    );
  }
}
