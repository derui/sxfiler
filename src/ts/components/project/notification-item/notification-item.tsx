import * as React from "react";
import { CSSTransition } from "react-transition-group";
import { Level, MessageBody } from "../../../domains/notification";

import * as ListItem from "../../ui/list-item/list-item";

// tslint:disable-next-line
const styles: ClassNames = require("./notification-item.module.scss");

interface ClassNames {
  root: string;
  animationExit: string;
  animationExitActive: string;
  animationEnter: string;
  animationEnterActive: string;
}

export interface Props {
  body: MessageBody;
  level: Level;
  timeouted: boolean;
  onExited: () => void;
}

export class Component extends React.PureComponent<Props> {
  constructor(props: Props) {
    super(props);
  }

  public render() {
    const { body, level } = this.props;

    const renderer = () => {
      let levelStyle = "";
      switch (level) {
        case Level.Info:
          levelStyle = "info";
          break;
        case Level.Warning:
          levelStyle = "warning";
          break;
        case Level.Error:
          levelStyle = "error";
          break;
      }

      return (
        <ListItem.Component className={styles.root} data-level={levelStyle}>
          {body.message}
        </ListItem.Component>
      );
    };

    return (
      <CSSTransition
        in={!this.props.timeouted}
        onExited={this.handleExited}
        timeout={200}
        unmountOnExit={true}
        classNames={{
          enter: styles.animationEnter,
          enterActive: styles.animationEnterActive,
          exit: styles.animationExit,
          exitActive: styles.animationExitActive,
        }}
      >
        {renderer}
      </CSSTransition>
    );
  }

  private handleExited = () => {
    this.props.onExited();
  };
}
