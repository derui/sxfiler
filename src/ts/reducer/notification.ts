// reducers for notification
import { Actions } from "../actions";
import ActionTypes from "../actions/types/notification";
import { empty, State } from "../types/store-state/notification";

export function reducer(state: State = empty(), action: Actions): State {
  switch (action.type) {
    case ActionTypes.timeout:
      return { ...state, notifications: state.notifications.remove(action.notificationId) };
  }
  return state;
}

export default reducer;
