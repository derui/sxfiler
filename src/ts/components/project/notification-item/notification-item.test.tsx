import { shallow } from "enzyme";
import * as React from "react";
import renderer from "react-test-renderer";

import * as Notification from "../../../domains/notification";

import { Component as T } from "./notification-item";

describe("Project", () => {
  describe("Notification Item", () => {
    it("should print correctly", () => {
      const notification = Notification.createMessage("id", Notification.Level.Info, "message");
      const onExited = () => {
        return;
      };

      const tree = renderer
        .create(
          <T body={notification.getMessageBody()} level={notification.level} onExited={onExited} timeouted={false} />
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should call timeout callback if timeouted and transition ended", done => {
      const notification = Notification.createMessage("id", Notification.Level.Info, "message");
      const onExited = () => {
        done();
      };

      const wrapper = shallow(
        <T body={notification.getMessageBody()} level={notification.level} onExited={onExited} timeouted={false} />
      );
      wrapper.simulate("exited");
    });

    it("should be able to render warning message", () => {
      const notification = Notification.createMessage("id", Notification.Level.Warning, "message");
      const onExited = () => {
        return;
      };

      const tree = renderer
        .create(
          <T body={notification.getMessageBody()} level={notification.level} onExited={onExited} timeouted={false} />
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
