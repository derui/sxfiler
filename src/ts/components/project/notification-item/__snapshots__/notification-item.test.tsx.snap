// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Notification Item should be able to render warning message 1`] = `
<div
  aria-selected={false}
  className="root"
  data-level="warning"
  role="listitem"
>
  message
</div>
`;

exports[`Project Notification Item should print correctly 1`] = `
<div
  aria-selected={false}
  className="root"
  data-level="info"
  role="listitem"
>
  message
</div>
`;
